### PR TITLE
Add support for apple.locales_to_exclude

### PR DESCRIPTION
**Pull request by chiragramani**
**Context**
We have a use-case where certain locales are only for testing our localization pipeline and they are totally disjoint from the locales that are a part of production-facing bundles. `locales_to_include` does a great job for debugging environments but needs a long list of accepted locales if we are building for production. There can also be a trivial maintenance cost associated. Hence, this PR adds support for `locales_to_exclude` so that we can explicitly filter out the testing locales.

### CODE REVIEW

File: apple/internal/processor.bzl [hunk 1]
```diff
@@ -317,6 +339,11 @@ def _bundle_partial_outputs_files(
                     # Skip files for locales that aren't in the locales for the base resources.
                     continue
 
+            if locales_to_exclude:
+                locale = bundle_paths.locale_for_path(parent_dir)
+                if locale and sets.contains(locales_to_exclude, locale):
+                    continue
```
Comment (mattrobmattrob) [inline apple/internal/processor.bzl:345]: Can this be put closer to the portion that does the `apple.locales_to_include` handling or is it not possible?

https://github.com/bazelbuild/rules_apple/blob/2c5e786fbcef8430d48c66c04b249937e4c8bcf2/apple/internal/partials/resources.bzl#L233-L235

And:

https://github.com/bazelbuild/rules_apple/blob/2c5e786fbcef8430d48c66c04b249937e4c8bcf2/apple/internal/partials/resources.bzl#L240-L246

Comment (chiragramani) [reply apple/internal/processor.bzl]: oh yeah, that definitely will be a better place! thanks, updated!

File: doc/common_info.md [hunk 2]
```diff
@@ -279,6 +279,13 @@ the exact names of the locales to be included.
 This can be used to improve compile/debug/test cycles because most developers
 only work/test in one language.
 
+#### Explicitly Excluding Locales
+
+Use `--define "apple.locales_to_exclude=foo,bar,bam"` where `foo,bar,bam` are
+the exact names of the locales to be excluded.
+
+Note: this overrides the `--define "apple.locales_to_include=foo,bar,bam"` i.e. if `foo` is present in `locales_to_include` but also present in `locales_to_exclude`, it will be excluded from the output bundle.
```
[the remark this answers is not in the document]
Comment (chiragramani) [reply doc/common_info.md:287]: Whenever `locales_to_include` is not defined properly,  a warning(debug print info) is thrown. I mainly updated it to keep it consistent, don't have strong opinions here. If you suggest, we should fail instead, will be happy to update the code. Let me know what you think!

File: test/macos_application_resources_test.sh [hunk 3]
```diff
@@ -149,4 +149,51 @@ EOF
       "app.app/Contents/Resources/bundle_library_macos.bundle/fr.lproj/localized.strings"
 }
 
+# Tests that the localizations that are explictly excluded via 
+# --define "apple.locales_to_exclude=fr" are not included in the output bundle.
+function test_bundle_localization_excludes() {
```
[the remark this answers is not in the document]
Comment (chiragramani) [reply test/macos_application_resources_test.sh:154]: sounds good, better code coverage :) updated

File: test/macos_application_resources_test.sh [hunk 4]
```diff
@@ -149,4 +149,83 @@ EOF
...
+
+  # Verify the app has a `it` localization and not an `fr` localization.
+  assert_zip_not_contains "test-bin/app/app.zip" \
+      "app.app/Contents/Resources/fr.lproj/localized.strings"
+  assert_zip_contains "test-bin/app/app.zip" \
+      "app.app/Contents/Resources/it.lproj/localized.strings"
+}
+
+# Tests that the localizations that are explictly excluded via 
+# --define "apple.locales_to_exclude=fr" overrides the ones explicitly included via "apple.locales_to_include" and are not included in the output bundle.
```
Comment (mattrobmattrob) [inline test/macos_application_resources_test.sh:192]: Mind line wrapping this a bit?

File: apple/internal/partials/resources.bzl [hunk 5]
```diff
@@ -231,6 +258,7 @@ def _resources_partial_impl(
     infoplists = []
 
     locales_requested = _locales_requested(config_vars = platform_prerequisites.config_vars)
+    locales_to_exclude = _locales_to_exclude(config_vars = platform_prerequisites.config_vars)
```
Comment (mattrobmattrob) [inline apple/internal/partials/resources.bzl:261]: Mind calling these variables and the function "locales_excluded" to match the inclusion variable and function name?

File: apple/internal/partials/resources.bzl [hunk 6]
```diff
@@ -244,6 +272,11 @@ def _resources_partial_impl(
                 elif locale != None:
                     sets.insert(locales_dropped, locale)
                     continue
+            if locales_to_exclude:
+                locale = bundle_paths.locale_for_path(parent_dir)
+                if locale and sets.contains(locales_to_exclude, locale):
```
Comment (mattrobmattrob) [inline apple/internal/partials/resources.bzl:277]: Do you need the `if locale` portion here or can it just be the set check like in the inclusion case?

File: apple/internal/partials/resources.bzl [hunk 7]
```diff
@@ -142,6 +142,33 @@ def _validate_processed_locales(*, label, locales_dropped, locales_included, loc
                   sets.str(unused_locales) + " in locale filter. Please verify " +
                   "apple.locales_to_include is defined properly.")
 
+        if locales_requested and locales_to_exclude:
+            conflicting_locales = sets.intersection(locales_requested, locales_to_exclude)
+            if sets.length(conflicting_locales):
+                # There is no way to issue a warning, so print is the only way
+                # to message.
+                # buildifier: disable=print
+                print("Warning: " + str(label) + " dropping " +
```
Comment (mattrobmattrob) [inline apple/internal/partials/resources.bzl:151]: @keith, do we have a precedent for this in the codebase?

**Review by mattrobmattrob — APPROVED** [review]

File: apple/internal/partials/resources.bzl (hunk 7 above)
Comment (keith) [reply apple/internal/partials/resources.bzl:151]: seems like we should just fail?

File: apple/internal/partials/resources.bzl (hunk 5 above)
Comment (chiragramani) [reply apple/internal/partials/resources.bzl]: sounds good!

File: apple/internal/partials/resources.bzl (hunk 6 above)
Comment (chiragramani) [reply apple/internal/partials/resources.bzl]: Yes, as `None` is the default return value of `_locales_excluded` and `sets` operations need a non-none value.

**Review by mattrobmattrob — APPROVED** [review]

**Pull request merged**